### ISSUE TITLE
Unexpected timeouts in Channel.CloseAsync in NetTcp Async Duplex scenario

### ISSUE THREAD

**Issue by KKhurin** [dotnet/wcf#1239]
Stress caught it after implementing a simple CloseAsync timing tracker.
The tail of the only GC root looks like this:

```
            ->  000000406207fc50 System.ServiceModel.Channels.IdlingCommunicationPool`2+IdleTimeoutEndpointConnectionPool[[System.String, System.Private.CoreLib],[System.ServiceModel.Channels.IConnection, System.Private.ServiceModel]]
            ->  000000406207fc90 System.Collections.Generic.List`1[[System.ServiceModel.Channels.IConnection, System.Private.ServiceModel]]
            ->  00000040620ec4c0 System.ServiceModel.Channels.IConnection[]
            ->  00000040621e3b78 System.ServiceModel.Channels.BufferedConnection
            ->  00000040621e3a98 System.ServiceModel.Channels.CoreClrSocketConnection
            ->  00000040622f7e18 System.Threading.Tasks.TaskCompletionSource`1[[System.Boolean, System.Private.CoreLib]]
            ->  00000040622f7e30 System.Threading.Tasks.Task`1[[System.Boolean, System.Private.CoreLib]]
            ->  00000040622f7ec0 System.Action
            ->  00000040622f7ea8 System.Runtime.CompilerServices.AsyncMethodBuilderCore+MoveNextRunner
            ->  00000040622f7f50 System.ServiceModel.Channels.SessionConnectionReader+<ReceiveAsync>d__27
            ->  000000406225b1b8 System.ServiceModel.Channels.ClientDuplexConnectionReader
            ->  000000406241e518 System.ServiceModel.Channels.ClientFramingDuplexSessionChannel
            ->  000000406241ed30 System.EventHandler
            ->  000000406241ed08 System.Object[]
            ->  000000406241eb88 System.EventHandler
            ->  000000406241eb28 System.ServiceModel.Dispatcher.DuplexChannelBinder
            ->  000000406241ed88 System.ServiceModel.Dispatcher.ChannelHandler
            ->  000000406241ebc8 System.ServiceModel.Channels.ServiceChannel
```

**Comment by zhenlan** [comment]
@roncain,  can you take a look to see if you can spot anything?

**Comment by roncain** [comment]
I investigated and am inclined to close this or allow @KKhurin to update the stress test as described below.

The stress tests for this sets its own timer at 50 seconds and the CloseTimeout for 60 seconds when it does a close async of the channel factory (under high loads).  If the 50 second time fires first, it declares an issue and breaks to the debugger.  But examining the DMP files indicated nothing appeared to be wrong.  I modified the stress test so that if the 50 second timer fired first, it just wrote to the Console and then did a Task.WhenAny waiting on the close async task and a new 50 second delay task.  At the end of that wait, if the close async task was not completed, it declared a problem and broke to the debugger.

I allowed this modified test to run for 16 hours on 2 separate occasions.  Roughly once per hour, we would see a close async task take longer than 50 seconds (CPU load was 100% on the client).  But in all cases, the close async task *did* complete before the 2nd 50 second timeout.  And in all cases, they were in the faulted state.

The fault was almost always a ProtocolException that had occurred when an incoming duplex message was received after its corresponding channel had been closed.  To put this in perspective, the stress test was actively closing channel factories while duplex operations were in-flight, so it is expected to see this kind of exception.  But in all cases, this exception correctly propagated back to the channel factory close async Task.

Incidentally, I ran this same test where the client used the full .NET framework.  Though it did not fail in exactly the same way, it did frequently take > 50 seconds for a duplex operation to complete (CPU load 100%).   I believe this is an indication that at these loads, 50 seconds is right on the edge of being too short a timeout for some of the duplex operations to complete.

I recommend @KKhurin consider modifying the stress tests to allow this 2-stage timeout approach and letting it run for a long time on the 24 core lab machines to see if another issue can be found.
